torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1362,6 +1362,160 @@ def rbinary_magic_impl(self, other):
+                    denominator_list.append(sympy.Pow(item.base, -item.exp))
+                else:
+                    numerator_list.append(item)
+
+            numerator = self._int(1) if len(numerator_list) == 0 else math.prod(self(numerator_list))
+
+            if len(denominator_list) > 0:
+                denominator = math.prod(self(denominator_list))
+                return self._div(numerator, denominator)
+
+            return numerator
+
+        def _Pow(self, expr: sympy.Pow):
+            base, exp = self(expr.args)
+            return base ** exp
+
+        def _FloorDiv(self, expr: FloorDiv):
+            base, divisor = self(expr.args)
+            return base / divisor
+
+        def __call__(self, expr):
+            if isinstance(expr, (tuple, list)):
+                return type(expr)(self(e) for e in expr)
+            if isinstance(expr, sympy.Rel):
+                typename = "Rel"
+            elif isinstance(expr, sympy.Integer):
+                typename = "int"
+                expr = int(expr)
+            elif isinstance(expr, sympy.core.numbers.IntegerConstant):
+                constantmap = {
+                    sympy.S.Zero: 0,
+                    sympy.S.One: 1,
+                    sympy.S.NegativeOne: -1,
+                }
+                typename = "int"
+                expr = constantmap[expr]
+            else:
+                typename = type(expr).__name__
+
+            method = f"_{typename}"
+            assert hasattr(self, method), (
+                f"unsupported sympy operation: {expr} ({method})"
+            )
+            return getattr(self, method)(expr)
+
+    class TranslationValidator:
+        def __init__(self) -> None:
+            super().__init__()
+            self._symbols = {}
+            self._inputs = set()
+            self._outputs = set()
+            self._z3 = SympyToZ3(self._symbols)
+
+        def _add_freesymbols(self, e: sympy.Expr) -> None:
+            for s in e.free_symbols:
+                if s not in self._symbols:
+                    assert isinstance(s, sympy.Symbol) and s.is_integer and s.is_positive
+                    self._symbols[s] = z3.Int(s.name)
+                    self._outputs.add(self._symbols[s] >= 0)
+
+        def _sympy_to_z3(self, expr: sympy.Expr):
+            z3_expr = self._z3(expr)
+            return z3_expr
+
+        def add_input(self, e: sympy.Expr) -> None:
+            self._add_freesymbols(e)
+            self._inputs.add(self._sympy_to_z3(e))
+
+        def add_output(self, e: sympy.Expr) -> None:
+            self._add_freesymbols(e)
+            self._outputs.add(self._sympy_to_z3(e))
+
+        def validate(self) -> bool:
+            from torch._dynamo.utils import dynamo_timed
+
+            if len(self._inputs) == 0 or len(self._outputs) == 0:
+                return True
+
+            solver = z3.Solver()
+            solver.add(z3.And(z3.Not(z3.And(*self._inputs)), z3.And(*self._outputs)))
```

Comment:
Actually it looks like you did it right here: give me a solution where outputs evaluates true (guard is let through) but inputs evaluates false.

Reply:
is this the equivalence check?
then this is what you want: And(inputs) != And(outputs)

Reply:
> is this the equivalence check?

It's not an equivalence check. But, [as Ed pointed out](https://github.com/pytorch/pytorch/pull/101146#issuecomment-1543193845), a soundness check: whether the outputs only yield true when the inputs yield true.

Reply:
By the way, I added comments. Hopefully, that's a bit clearer now.

Reply:
> whether the outputs only yield true when the inputs yield true

If you wanted to prove that input => output, counterexamples would satisfy input = true, output = false...which is NOT what the condition above is asking for.

Instead, I think what we're really trying to prove is that output => input. In other words, we are fine with stricter outputs given relaxed inputs (e.g., when input is x > 5 and output is x > 6, there should be no counterexample). This matches what Ed said and what the condition above is asking for. In other words, the code is correct (but perhaps the dialog is confusing).

Reply:
> In other words, it tries to find an assignment to the free variables such that the conjunction of the initial expressions (input expressions) yields TRUE, but the conjunction of the final expressions (output expressions -- those actually used for building the guard function) yields FALSE.

The PR summary also has the same bug. It's the opposite of what the code is doing.

Reply:
> In other words, the code is correct (but perhaps the dialog is confusing).

I guess I tried to express the correct thing with the word _"only"_ there. But I see that that is confusing. I will try to reword that.

> The PR summary also has the same bug.

Ah, yes. It's the other way around in the PR description.